src/DebuggerForm.cpp

```diff
@@ -1535,3 +1505,33 @@ void DebuggerForm::addressSlot(int addr, qint8& ps, qint8& ss, qint16& segment)
 			segment = memLayout.romBlock[q];
 	}
 }
+
+void DebuggerForm::reloadBreakpoints(bool merge)
+{
+	auto command = new Command("debug_list_all_breaks",
+		[this, merge](const QString& message){ emit replyReceived(message, merge); }
+	);
+	comm.sendCommand(command);
+}
+
+void DebuggerForm::processBreakpoints(const QString& message, bool merge)
+{
+	if (merge) {
+		QString bps = session.breakpoints().mergeBreakpoints(message);
+		if (!bps.isEmpty()) {
+			comm.sendCommand(new SimpleCommand(bps));
+			reloadBreakpoints(false);
+		} else {
+			disasmView->update();
+			session.sessionModified();
+			updateWindowTitle();
+		}
+		bpView->sync();
+	} else {
+		session.breakpoints().setBreakpoints(message);
+		disasmView->update();
+		session.sessionModified();
+		updateWindowTitle();
+		bpView->sync();
+	}
+}
```

Comment:
This function seems a bit weird to me. I'm going to make two suggestions, but I haven't fully worked them, so I don't know if they will actually improve the code. Please consider them, but also fine if in the end you decide to stick with the current structure.

* Depending on the `merge` parameter, this function does two completely different things. Whenever I see a function like that I wonder if it isn't better to split this function into two functions (without boolean parameter), and then call one or the other depending on the original parameter.
Applied to this specific situation that would mean in `reloadBreakPoints()` register one of these two (splitted) functions as the `okCallback`  (I'm assuming we remove the signal/slot stuff).
Or better: take this one step further and split `reloadBreakpoints()` into two: e.g. `mergeBreakpoints()` and `reloadBreakpoints()`.

* The 2 else-blocks (with the GUI update stuff) are largely duplicated. I wonder if it isn't possible to remove this duplication?
Maybe, if you apply the split from the previous point, then in `processMerge()` you could always call `reloadBreakpoints()` (also if `bps` is empty). And then only do the GUI update stuff in `processBreakpoints()`? (This does a bit too much work on merge when there were no breakpoints defined yet in openMSX, but this is anyway an infrequent operation).

Reply:
I don't like it either. It is a mess that is hard to understand.